cmd/formatter/shortcut.go

```diff
@@ -0,0 +1,324 @@
+func (lk *LogKeyboard) PrintKeyboardInfo(printFn func()) {
+	lk.clearNavigationMenu()
+	printFn()
+
+	if lk.logLevel == INFO {
+		lk.createBuffer(0)
+		lk.printNavigationMenu()
+	}
+}
+
+// Creates space to print error and menu string
+func (lk *LogKeyboard) createBuffer(lines int) {
+	allocateSpace(lines)
+
+	if lk.kError.shoudlDisplay() {
+		extraLines := linesOffset(lk.kError.error()) + 1
+		allocateSpace(extraLines)
+		lines += extraLines
+	}
+
+	infoMessage := lk.navigationMenu()
+	extraLines := linesOffset(infoMessage) + 1
+	allocateSpace(extraLines)
+	lines += extraLines
+
+	if lines > 0 {
+		MoveCursorUp(lines)
+	}
+}
+
+func (lk *LogKeyboard) printNavigationMenu() {
+	if lk.logLevel == INFO {
+		height := goterm.Height()
+		menu := lk.navigationMenu()
+
+		MoveCursorX(0)
+		SaveCursor()
+
+		lk.kError.printError(height, menu)
+
+		MoveCursor(height-linesOffset(menu), 0)
+		ClearLine()
+		fmt.Print(menu)
+
+		MoveCursorX(0)
+		RestoreCursor()
+	}
+}
+
+func (lk *LogKeyboard) navigationMenu() string {
+	var options string
+	var openDDInfo string
+	if lk.IsDockerDesktopActive {
+		openDDInfo = shortcutKeyColor("v") + navColor(" View in Docker Desktop")
+	}
+	var watchInfo string
+	if openDDInfo != "" {
+		watchInfo = navColor("   ")
+	}
+	var isEnabled = " Enable"
+	if lk.Watch.Watching {
+		isEnabled = " Disable"
+	}
+	watchInfo = watchInfo + shortcutKeyColor("w") + navColor(isEnabled+" Watch")
+	return options + openDDInfo + watchInfo
+}
+
+func (lk *LogKeyboard) clearNavigationMenu() {
+	height := goterm.Height()
+	MoveCursorX(0)
+	SaveCursor()
+	for i := 0; i < height; i++ {
+		MoveCursorDown(1)
+		ClearLine()
+	}
+	RestoreCursor()
+}
+
+func (lk *LogKeyboard) PrintEnter() {
+	lk.clearNavigationMenu()
+	lk.printNavigationMenu()
+}
+
+func (lk *LogKeyboard) CleanTerminal() {
+	height := goterm.Height()
+	for i := 0; i < height; i++ {
+		NewLine()
+		ClearLine()
+	}
+}
+
+func (lk *LogKeyboard) openDockerDesktop(ctx context.Context, project *types.Project) {
+	if !lk.IsDockerDesktopActive {
+		return
+	}
+	eg.Go(tracing.EventWrapFuncForErrGroup(ctx, "menu/gui", tracing.SpanOptions{},
+		func(ctx context.Context) error {
+			link := fmt.Sprintf("docker-desktop://dashboard/apps/%s", project.Name)
+			err := open.Run(link)
+			if err != nil {
+				err = fmt.Errorf("Could not open Docker Desktop")
+				lk.kError.addError("View", err)
+			}
+			return err
+		}),
+	)
+}
+
+func (lk *LogKeyboard) StartWatch(ctx context.Context, project *types.Project, options api.UpOptions) {
+	if !lk.IsWatchConfigured {
+		eg.Go(tracing.EventWrapFuncForErrGroup(ctx, "menu/watch", tracing.SpanOptions{},
+			func(ctx context.Context) error {
+				err := fmt.Errorf("Watch is not yet configured. Learn more: %s", ansiColor(CYAN, "https://docs.docker.com/compose/file-watch/"))
+				lk.kError.addError("Watch", err)
+				return err
+			}))
+		return
+	}
+	lk.Watch.switchWatching()
+	if !lk.Watch.isWatching() && lk.Watch.Cancel != nil {
+		lk.Watch.Cancel()
+	} else {
```

Comment:
should this be 

```go
	lk.Watch.switchWatching()
	if !lk.Watch.isWatching() {
		if lk.Watch.Cancel != nil {
			lk.Watch.Cancel()
		}
	} else {
```

(that is, even if the cancel function is nil, we only want to start watch if `isWatching == true`, right?)

Reply:
yes, I think this was an old bug that was fixed, I think it was because the configuration validation was being done after and not before at some point. Both expressions are equivalent

Reply:
nice catch